### ISSUE TITLE
"How to get all nodes within ellipse" and higher dimensions

### ISSUE THREAD

**Issue by Popkultur** [bayespy/bayespy#89]
I am sorry, I am not a mathematician and not fluent in english, so I hardly understand what's going on in these algorythms. I appreciate your work on this module. I managed to implement it for a test run on my own data. This data is in at least 3 dimensions, so I would really like to know how to implement the gaussian mixture model in higher dimensional data like 3D or 4D. I also would like to know how to visualize that (maybe with sort of a function gaussian_mixture_3d and mplot3d.)

For now, it would help me the most if I knew how to get all nodes within the radius of an ellipse in the gaussian mixture model. If this would be or is already possible for higher dimensions than 2, I'd like to know either.

What I try to do is to detect accumulations of nodes in higher dimensional (3d at least) data.

**Comment by jluttine** [comment]
A Gaussian mixture model example can be found here:
http://bayespy.org/en/latest/examples/gmm.html

There's an explicit variable `D` which is the dimensionality of the data. In the example it's 2, but you can make it 3 or 4 or whatever.

Visualizing high-dimensional data is a problem of its own. 3D data would work to some extent with mplot3d but I haven't implemented anything for that yet. For any dimensional data, you can always examine the cluster assignments of your data points with `bayespy.plot.hinton(Z)` or by just printing `print(Z.get_moments()[0])`. You could also find the most probable clusters for each data point with `i = np.argmax(Z.get_moments()[0], axis=-1)` and then code the cluster indices `i` into unique colors for your 3D scatter plot of data.

Finding data points that are, for instance, within the one standard deviation ellipse of a Gaussian component is slightly more involved but certainly doable. Note that it is possible that some points are not inside any ellipse and some points are inside multiple ellipses. Perhaps much easier would be to find data points that have a probability density above some threshold if just considering the plain Gaussian density function of a cluster. That takes into account how the density changes when variance changes. And that's roughly the measure the mixture model uses for assigning clusters.

**Comment by Popkultur** [comment]
Thank you!

> Perhaps much easier would be to find data points that have a probability density above some threshold if just considering the plain Gaussian density function of a cluster.

Yes, this would be indeed the best option. How to archive that? I tried the Gaussian pdf or get_pdf_nodes() since the command mu = Gaussian(np.zeros(D), 1e-5*np.identity(D), plates=(K,), name='mu') returns a Gaussian object, but it returns a onedimensional array for pdf (probably not what I need) and some sort of Mixture object for the latter. I thought I could get an array of objects which contain every nodes Gaussian density.

**Comment by jluttine** [comment]
For instance, this should give you the *log* probability density function of the data points at each Gaussian cluster: `Gaussian(mu, Lambda).logpdf(y[:,None,:])`.

Note that this doesn't take into account cluster assignments prior probabilities. If you want to use those too, then you can just directly use `Z.get_moments()[0]` to get the posterior cluster assignment probabilities.

**Comment by Popkultur** [comment]
```
    y = mu.pdf(x[:,None,:])
    min_max_scaler = preprocessing.MinMaxScaler()
    X_train_minmax = min_max_scaler.fit_transform(y)
    #X_train_minmax = np.append(np.vstack(np.ones(len(X_train_minmax))),X_train_minmax,1)   
    mycmap = matplotlib.cm.brg 
    ax.scatter(x[:,0], x[:,1], x[:,2], alpha=1, c=range(len(X_train_minmax)), cmap=mycmap) 
```
this is what I tried (where x is my 3d data and y is the 4th dimension). However, it does not color correctly by density.

**Comment by jluttine** [comment]
Not `mu.pdf(x[:,None,:])`, but `Gaussian(mu, Lambda).logpdf(y[:,None,:])` gives you the density for data points `y` for every Gaussian cluster with means `mu` and precisions `Lambda`.
